### PR TITLE
Filter out the user provided unsafe packages

### LINKED ISSUE #1756: unsafe-package is not overriding the default list of packages
```
$ pip-compile --help
  --unsafe-package TEXT           Specify a package to consider unsafe; may be
                                  used more than once. Replaces default unsafe
                                  packages: distribute, pip, setuptools
```

#### Environment Versions

1. OS Type `macos 13.0.1`
1. Python version: `3.9.7`
1. pip version: `22.3.1`
1. pip-tools version: `6.11.0`

#### Steps to replicate

1. define `foo.in`

```
pyramid
```

2. `pip-compile --unsafe-package pip foo.in`

#### Expected result

Pinned setuptools.

#### Actual result

Unpinned setuptools.

Below are the commands with and without `--unsafe-package` and note the output is the same.

```
❯ pip-compile foo.in
WARNING: using legacy resolver is deprecated and will be removed in future versions. The default resolver will be change to 'backtracking' in 7.0.0 version. Specify --resolver=backtracking to silence this warning.
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile foo.in
#
hupper==1.10.3
    # via pyramid
pastedeploy==3.0.1
    # via plaster-pastedeploy
plaster==1.1.2
    # via
    #   plaster-pastedeploy
    #   pyramid
plaster-pastedeploy==1.0.1
    # via pyramid
pyramid==2.0
    # via -r foo.in
translationstring==1.4
    # via pyramid
venusian==3.0.0
    # via pyramid
webob==1.8.7
    # via pyramid
zope-deprecation==4.4.0
    # via pyramid
zope-interface==5.5.2
    # via pyramid

# The following packages are considered to be unsafe in a requirements file:
# setuptools
```

```
❯ pip-compile --unsafe-package pip foo.in
WARNING: using legacy resolver is deprecated and will be removed in future versions. The default resolver will be change to 'backtracking' in 7.0.0 version. Specify --resolver=backtracking to silence this warning.
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --unsafe-package=pip foo.in
#
hupper==1.10.3
    # via pyramid
pastedeploy==3.0.1
    # via plaster-pastedeploy
plaster==1.1.2
    # via
    #   plaster-pastedeploy
    #   pyramid
plaster-pastedeploy==1.0.1
    # via pyramid
pyramid==2.0
    # via -r foo.in
translationstring==1.4
    # via pyramid
venusian==3.0.0
    # via pyramid
webob==1.8.7
    # via pyramid
zope-deprecation==4.4.0
    # via pyramid
zope-interface==5.5.2
    # via pyramid

# The following packages are considered to be unsafe in a requirements file:
# setuptools
```

Note that if I use `--allow-unsafe` then everything ends up pinned, including pip in my env where pip gets included as well. Specifically pip-tools itself depends on pip so and we have pip-tools in one of our requirements files.

Also note that it says Python 3.10 was used which is incorrect. Opening a separate issue for this.

### PR DESCRIPTION
**Pull request by q0w**
Closes #1756 
##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [x] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by q0w** [comment]
Does unsafe_requirements in writer._iter_lines provide None as default value only for tests? Because reolver.unsafe_requirements is an empty set by default.

**Comment by atugushev** [comment]
> Does unsafe_requirements in writer._iter_lines provide None as default value only for tests? 

It looks like it does.

**Review by atugushev — APPROVED** [review]
Awesome! 🎉

**Comment by q0w** [comment]
Should I refactor unsafe_requirements and unsafe_packages without None in _iter_lines?

File: tests/test_writer.py [hunk 1]
```diff
@@ -389,16 +406,23 @@ def test_write_order(writer, from_line):
...
     ]
-    assert list(writer._iter_lines(packages)) == expected_lines
+    assert (
+        list(
+            writer._iter_lines(
+                packages, unsafe_requirements=set(), unsafe_packages=set(), markers={}
+            )
+        )
+        == expected_lines
+    )
```
Comment (atugushev) [inline tests/test_writer.py:428]:
```suggestion
    lines = list(
        writer._iter_lines(
            packages, unsafe_requirements=set(), unsafe_packages=set(), markers={}
        )
    )
    assert lines == expected_lines

```


Maybe add an intermediate variable? This looks more readable to me. What do you think?

Comment (q0w) [reply tests/test_writer.py]: done

**Comment by q0w** [comment]
@atugushev could you enable auto-merge again, i pushed again

**Review by ssbarnea — APPROVED** [review]

**Comment by atugushev** [comment]
Thanks @q0w for fixing this!

**Pull request merged**